client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java

```diff
@@ -0,0 +1,666 @@
+         // TODO: extract async threads better
+         int asyncThreads = new ConfigurationProperties(configuration.asyncExecutorFactory().properties()).getDefaultExecutorFactoryPoolSize();
+         // static field with default is private in MultithreadEventLoopGroup
+         int eventLoopThreads = SecurityActions.getIntProperty("io.netty.eventLoopThreads", Runtime.getRuntime().availableProcessors() * 2);
+         // Note that each event loop opens a selector which counts
+         int maxExecutors = Math.min(asyncThreads, eventLoopThreads);
+         this.eventLoopGroup = new NioEventLoopGroup(maxExecutors, executorService);
+
+         Collection<SocketAddress> servers = new ArrayList<>();
+         initialServers = new ArrayList<>();
+         for (ServerConfiguration server : configuration.servers()) {
+            servers.add(InetSocketAddress.createUnresolved(server.host(), server.port()));
+         }
+         initialServers.addAll(servers);
+         if (!configuration.clusters().isEmpty()) {
+            configuration.clusters().forEach(cluster -> {
+               Collection<SocketAddress> clusterAddresses = cluster.getCluster().stream()
+                       .map(server -> InetSocketAddress.createUnresolved(server.host(), server.port()))
+                       .collect(Collectors.toList());
+               ClusterInfo clusterInfo = new ClusterInfo(cluster.getClusterName(), clusterAddresses);
+               log.debugf("Add secondary cluster: %s", clusterInfo);
+               clusters.add(clusterInfo);
+            });
+            clusters.add(new ClusterInfo(DEFAULT_CLUSTER_NAME, initialServers));
+         }
+         currentClusterName = DEFAULT_CLUSTER_NAME;
+         topologyInfo = new TopologyInfo(defaultCacheTopologyId, Collections.unmodifiableCollection(servers), configuration);
+         operationsFactory = new OperationsFactory(this, codec, configuration);
+         maxRetries = configuration.maxRetries();
+
+         if (log.isDebugEnabled()) {
+            log.debugf("Statically configured servers: %s", servers);
+            log.debugf("Load balancer class: %s", configuration.balancingStrategyClass().getName());
+            log.debugf("Tcp no delay = %b; client socket timeout = %d ms; connect timeout = %d ms",
+                    configuration.tcpNoDelay(), configuration.socketTimeout(), configuration.connectionTimeout());
+         }
+         balancers = new HashMap<>();
+         WrappedByteArray defaultCacheName = new WrappedByteArray(RemoteCacheManager.cacheNameBytes());
+         balancers.put(defaultCacheName, createBalancer(defaultCacheName));
+      } finally {
+         lock.writeLock().unlock();
+      }
+      pingServersIgnoreException();
+   }
+
+   private ChannelPool newPool(SocketAddress address) {
+      log.debugf("Creating new channel pool for %s", address);
+      Bootstrap bootstrap = new Bootstrap()
+            .group(eventLoopGroup)
+            .channel(NioSocketChannel.class)
```

Comment:
We could probably use Epoll here (if it is supported). See how this is done in Hot Rod Server.

Reply:
Yes, I've considered this but committed the smallest possible set of changes. Would you mind adding that as separate PR, or should I include that in this one?

Reply:
Either way works for me. But I guess Epoll will have impact on performance so I would include it here if I were in your shoes.

Reply:
Ok, no issues with that.